conda-store-server/conda_store_server/orm.py

```diff
@@ -225,15 +233,29 @@ def build_key(self):
         The build key should be a key that allows for the environment
         build to be easily identified and found in the database.
         """
-        datetime_format = "%Y%m%d-%H%M%S-%f"
-        return f"{self.specification.sha256}-{self.scheduled_on.strftime(datetime_format)}-{self.id}-{self.specification.name}"
+        if self.build_key_version == 1:
+            datetime_format = "%Y%m%d-%H%M%S-%f"
+            return f"{self.specification.sha256}-{self.scheduled_on.strftime(datetime_format)}-{self.id}-{self.specification.name}"
+        elif self.build_key_version == 2:
+            hash = self.specification.sha256[:4]
+            timestamp = int(self.scheduled_on.timestamp())
+            id = self.id
+            name = self.specification.name[:16]
+            return f"{hash}-{timestamp}-{id}-{name}"
```

Comment:
These lengths (4 and 16) are hardcoded and there's no explanation on why that was chosen. Instead I'd rather have a fixed-length hash à la nix.

Reply:
I like Nix hashes as well (specifically, I'm talking about nix-store hashes, since those also use the filesystem, see [here](https://nixos.org/guides/nix-pills/nix-store-paths)), but it can happen later and in a separate PR. It requires quite a bit of work and it was pointed out to me during the last team meeting that we might want to allocate our time elsewhere for now.

Here are some questions to answer for Nix hashes:
- To make this really fixed-size, we would need to design the v2 path scheme such that it also doesn't include the namespace name. It should be just: `store_directory / hash / <env contents>` where the first two components is the prefix that needs to be <= 255 to satisfy the conda prefix contraint.
- To hash like this and avoid weird issues due to mismatched version, we cannot hash the specification as we do now, we need to hash the contents of the lock file. Specs don't have the versions pinned, lock files are deterministic.
- Since the hashes are shared between users (no namespace anymore), they need to be resistant to collisions. The size of the hash and its collision resistance are related. This can be estimated to find the best size for this use case.
- Hashes are one-way (by design). But we also use `build_key`s to get the `build_id`. See `parse_build_key` and how it's used in `get_docker_image_manifest`. For that, we'd need to store an extra `hash -> id` mapping in the DB.
- IIUC, all other parts of the `build_key` are just to help with debugging or search. You can verify which env on disk corresponds to which specification, see when it was built, and get a human-readable name of the env.
- `build_id`s are unique because they are primary keys in the `Build` table. The only concern here is that they might potentially get reused by some DB backends.

How 4 and 16 were selected:
- Because only the `build_id`s matter, everything else doesn't help with the uniqueness part, just helps with debugging
- I selected 4 to still be able to verify that a spec matches the env by hashing it and have some confidence
- I selected 16 because it contained enough information to be meaningful with some env names I've tried, but also because it's short enough.

Summary:
- Since #653 adds error checking for this and documentation, it makes it easier for users to understand what's going on
- I'm going to bump 4 to 8 since it's a more standard size for truncation (even though it's not necessary here). This makes it somewhat less arbitrary. The original proposal was a unix timestamp and a truncated hash.
- I'm going to remove the truncation from names (since the linked PR has instructions for users to use shorter names if there are problems). It removes this arbitrary decision, which might lead to confusion later.
- Let's discuss the Nix-hash-like idea separately in a new issue.

Reply:
> It requires quite a bit of work and it was pointed out to me during the last team meeting that we might want to allocate our time elsewhere for now.

My main concern here is that we are going to break the path interface already, so maybe there's a benefit in only doing it once and for good. At the same time, if we perfect the art of migrating paths, then it's ok :) I guess it can be discussed in the weekly tomorrow.

Reply:
There's no breakage. Everything is migrated transparently and you can switch back and forth. Nix-like hashes would be more different compared to what we have right now. I still see value in having just a truncated hash and timestamp (as in the v2 here). Nix-style hashes can be added later.